### ISSUE TITLE
Kubectl can apply config files in bulk

### ISSUE THREAD

**Issue by smarterclayton** [GoogleCloudPlatform/kubernetes#1905]
Extracted from #1325

* Postponed `submit [<filename>...]` - This command exists to satisfy the second use case above. Reconcile and submit any changes from a given set of config(s), from files or from stdin. If you had an entire directory tree of files that had configs that represented your cluster state, you could use this command to submit them all to either create or update your cluster.
* Postponed `diff [<filename>...]` - A dry-run version of submit.

Also relevant #1007,  #113, #1702, #1704, [OpenShift config](https://github.com/openshift/origin/tree/master/pkg/config) and [OpenShift apply command](https://github.com/openshift/origin/blob/master/pkg/cmd/client/kubecfg.go#L519), and #987 which was the first attempt at this.  Also see [issues with config-deployment label](https://github.com/GoogleCloudPlatform/kubernetes/labels/area/config-deployment).

I'd like to be able to represent a set of potentially unknown config objects inside a JSON file that can be applied en-masse - we currently almost all the necessary support in the encode/decode paths to be able to take an api versioned list, and attempt to extract unknown objects as JSON and if necessary post them to an endpoint directly.

The client-flexible version of this is dependent on #1355 in order to be able to find the endpoints for resources that are not pod, controller, or service.

**Comment by brendandburns** [comment]
@smarterclayton @ghodss  @bgrant0607 

This is completed, correct?

**Comment by bgrant0607** [comment]
Not entirely, but remaining items are covered by cli-roadmap.md.

**Comment by JeanMertz** [comment]
I couldn't find this in the cli-roadmap.md document (or any of its linked GH issues), and wasn't able to see any mention of it in this issue, so I'll just put this here, but probably this has been mentioned somewhere already:

```
$ kubectl create -f rc-*.json
error: Unexpected args: [rc-2.json rc-3.json rc-4.json rc-5.json]
see 'kubectl create -h' for help.
```

```
$ kubectl create -f rc-1.json,rc-2.json,rc-3.json,rc-4.json,rc-5.json
replicationcontrollers/consul-ivake
replicationcontrollers/consul-8mtxa
replicationcontrollers/consul-wzb43
replicationcontrollers/consul-fppmw
replicationcontrollers/consul-au3wp
```

it would be nice if the first syntax (ie. array of files) was supported.

**Comment by smarterclayton** [comment]
Due to shell expansion it's tricky the way we currently have it.
Originally we wanted to allow args to the server and files in the same
command, so -f took an arg rather than modified the command.  In the future
we might want to allow you to create from a template on the server (pod
template, generic template, RC from a deployment, etc) so we should
probably be careful about expanding -f.  We could support another flag -F
that turns all arguments into files

On Sep 13, 2015, at 5:42 AM, Jean Mertz <notifications@github.com> wrote:

I couldn't find this in the cli-roadmap.md document (or any of its linked
GH issues), and wasn't able to see any mention of it in this issue, so I'll
just put this here, but probably this has been mentioned somewhere already:

$ kubectl create -f rc-*.json
error: Unexpected args: [rc-2.json rc-3.json rc-4.json rc-5.json]
see 'kubectl create -h' for help.

$ kubectl create -f rc-1.json,rc-2.json,rc-3.json,rc-4.json,rc-5.json
replicationcontrollers/consul-ivake
replicationcontrollers/consul-8mtxa
replicationcontrollers/consul-wzb43
replicationcontrollers/consul-fppmw
replicationcontrollers/consul-au3wp

it would be nice if the first syntax (ie. array of files) was supported.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/issues/1905#issuecomment-139856960>
.
